scheduler/core/components/optimizer/timeline.py

```diff
@@ -25,11 +26,109 @@ class Timeline:
     _total_time_slots: int
 
     def __post_init__(self):
-        self.schedule = np.full(self._total_time_slots, EMPTY)
+        self.time_slots = np.full(self._total_time_slots, EMPTY)
         self.is_full = False
 
     def __contains__(self, obs: Observation) -> bool:
-        return any(visit.obs_id == obs.id for visit in self.visits)
+        return any(obs.id in self.time_slots)
+
+    def _empty_slots(self) -> np.ndarray:
+        """
+        Determine the number of empty time slots..
+        """
+        return np.where(self.time_slots == EMPTY)[0][:]
+
+    def _available_intervals(self, empty_slots: np.ndarray) -> np.ndarray:
+        """
+        Calculate the available intervals in the schedule by creating an array that contains all
+        the groups of consecutive numbers
+        """
+        return np.split(empty_slots, np.where(np.diff(empty_slots) != 1)[0]+1)
```

Comment:
This is really nice... I didn't know this was possible in numpy! 👍

Reply:
This is from Sergio